### PR TITLE
feat(Optuna): Allow for parsing of Choice Nodes

### LINKED ISSUE #289: Support for conditionals in optimizers like Optuna
**Is your feature request related to a problem? Please describe.**
Right now the Optuna parser [does not support conditionals](https://github.com/automl/amltk/blob/b680838a5a4d5f97c6be0d9c481b8f66d9e2fca9/src/amltk/pipeline/parsers/optuna.py#L200). The TODO says that this functionality is not yet supported as we can't encode this into a static Optuna search space. Is this support infeasible or possible and on the roadmap?

**Describe the solution you'd like**
A way to still use Choice pipeline components and also Optuna. Ideally the resulting behaviour would be the same as when using Optuna and a Pythonic search space directly, as in the [Optuna docs](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html#defining-parameter-spaces).

**Describe alternatives you've considered**
Alternatively, you can create one component with a function that will call the choice list depending on a categorical variable, but then the other parameters are not as nicely passed only to the function where they make sense. So many configs in the space will be invalid.

### PR DESCRIPTION
**Pull request by berombau**
#### What does this implement/fix? Explain your changes.
Fixes #289.

#### Minimal Example / How should this PR be tested?

There are some extra tests, see `pytest -k optuna`.

#### Any other comments?

Seems to be working. Will add new issues if I run into issues with it. 

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/amltk/optimization/optimizers/optuna.py [hunk 1]
```diff
@@ -292,7 +292,36 @@ def ask(
...
+                if "__choice__" in name:
+                    possible_choices = distribution.choices
+                    choice_made = optuna_trial.suggest_categorical(name, choices=possible_choices)
+                    for c in possible_choices:
+                        if c != choice_made:
+                            delete_other_options.append(f"{name}:{c}:")
+            # filter all parameters given the made choices
+            filtered_workspace = {k: v for k, v in workspace.items() if (
+                ("__choice__" not in k) and
+                (not any(c in k for c in delete_other_options))
```
Comment (eddiebergman) [inline src/amltk/optimization/optimizers/optuna.py:310]: This line might need a second look, I had a bug a while back with something similar where essentially the `c` was present in some other `k` in another part of the pipeline that shouldn't have been effected by the given choice.

I'm on my phone so I cant write a concrete example but see the overall review comment for more

**Review by eddiebergman — CHANGES_REQUESTED** [review]

**Comment by eddiebergman** [comment]
Looks really good, clean and well done, many thanks!

I would ask that the actual sampling function is factored into its own function that is then tested a bit more as to it's sampling behaviour.

This wasn't required for the other optimizers as I relied on them testing them own functionality, i.e. I dont want to test that other optimizers have their internals correct.

However here, since we are the ones implementing the sampling procedure, might be good to have it factored out and testable.

You can make it a private function of the module outside of the class as generally I don't think a user should be exposed to it.

For the actual study object, you can choose any sampler that makes sense to you!

Things to test are what was mentioned in the comment but also maybe that deeper more heirarchical pipelines work too. You can see this file and likely just copy and paste one of those, or however you see fit!

**Comment by eddiebergman** [comment]
The random elements is usually just a thing of not giving a component a name. To do the configuring properly, I had to give them some identifier.

In any case, happy to push this through once the test pass. In any case, I can turn the remaining thing into an issue so this can be merged!

Very appreciative of your work!

**Comment by berombau** [comment]
Thanks for the help! It's a really nice toolkit and codebase, was very glad when I found it 👍 

The mypy crashes on my pre-commit, so running it locally.
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 2

tests/scheduling/test_scheduler.py:116: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.8.0
```

**Comment by eddiebergman** [comment]
> Thanks for the help! It's a really nice toolkit and codebase, was very glad when I found it 👍
> 
> The mypy crashes on my pre-commit, so running it locally.
> 
> ```
> mypy.....................................................................Failed
> - hook id: mypy
> - exit code: 2
> 
> tests/scheduling/test_scheduler.py:116: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
> https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
> Please report a bug at https://github.com/python/mypy/issues
> version: 1.8.0
> ```

Aye I'm aware of this issue, it's very annoying. I can't figure out what causes it but clearing cache seems to sometimes fix it -_-

**Pull request merged**